spec/amber/router/params_spec.cr

```diff
@@ -0,0 +1,81 @@
+    context "when parsing request body" do
+      context "when content-type is form url urlencoded" do
+        it "returns params" do
+          headers["Content-Type"] = "application/x-www-form-urlencoded"
+          body = "name=John Doe"
+          request = HTTP::Request.new("POST", "/", headers, body)
+          request.params["name"].should eq "John Doe"
+        end
+
+        it "parses body params with charset" do
+          headers["Content-Type"] = "application/x-www-form-urlencoded; charset=UTF-8"
+          body = "\x68\x65\x6C\x6C\x6F\x3D\x77\x6F\x72\x6C\x64"
+          request = HTTP::Request.new("POST", "/", headers, body)
+          request.params["hello"].should eq "world"
+        end
+      end
+
+      context "when parsing multipart" do
+        it "parses files from multipart forms" do
+          headers["Content-Type"] = "multipart/form-data; boundary=fhhRFLCazlkA0dX"
+          body = "--fhhRFLCazlkA0dX\r\nContent-Disposition: form-data; name=\"_csrf\"\r\n\r\nPcCFp4oKJ1g-hZ-P7-phg0alC51pz7Pl12r0ZOncgxI\r\n--fhhRFLCazlkA0dX\r\nContent-Disposition: form-data; name=\"title\"\r\n\r\ntitle field\r\n--fhhRFLCazlkA0dX\r\nContent-Disposition: form-data; name=\"picture\"; filename=\"index.html\"\r\nContent-Type: text/html\r\n\r\n<head></head><body>Hello World!</body>\r\n\r\n--fhhRFLCazlkA0dX\r\nContent-Disposition: form-data; name=\"content\"\r\n\r\nseriously\r\n--fhhRFLCazlkA0dX--"
+
+          request = HTTP::Request.new("POST", "/?test=example", headers, body)
+
+          request.params.files["picture"].filename.should eq "index.html"
+          request.params["title"].should eq "title field"
+          request.params["_csrf"].should eq "PcCFp4oKJ1g-hZ-P7-phg0alC51pz7Pl12r0ZOncgxI"
+        end
+      end
+
+      context "when parsing JSON body" do
+        it "parses json hash" do
+          headers["Content-Type"] = "application/json"
+          body = %({ "test": "test", "address": { "city": "New York" }})
+          request = HTTP::Request.new("POST", "/", headers, body)
+
+          request.params["test"].should eq "test"
+          request.params.json("address")["city"].should eq "New York"
+        end
+
+        it "parses json array" do
+          headers["Content-Type"] = "application/json"
+          body = %(["test", "test2"])
+          request = HTTP::Request.new("POST", "/", headers, body)
+
+          request.params.json("_json").should eq %w(test test2)
+        end
+      end
+
+      context "when parsing route params" do
+        it "parses params from route" do
+          handler = ->(context : HTTP::Server::Context) {}
+          route = Route.new("GET", "/fake/action/:id/:name", handler, :action, :web, "", "FakeController")
+          Amber::Server.router.add(route)
+          request = HTTP::Request.new("GET", "/fake/action/123/john")
+
+          request.route
+
+          request.params["id"].should eq "123"
+          request.params["name"].should eq "john"
+        end
+      end
+    end
+  end
```

Comment:
I think it would be good to add a test for how different parameters can override each other, so it's documented to not change accidentally.

eg, what happens when this route: `/objects/:id/:name` is hit with this url: `/objects/23/foo?name=bar`?

Reply:
Yeah, fetch all is fine, but I'm asking about the regular fetch.

What happens when you call `params["name"]` in a `/objects/:id/:name` action that was triggered with a url like this: `/objects/23/foo?name=bar`

I think there should be a test defining what happens when a parameter is shadowed, because I don't expect the majority of use cases to use list fetch with `fetch_all`.

Reply:
Apparently with `HTTP::Params` fetch returns the first one added rather than the last one. Where as fetch_all returns the array.

Reply:
Well, that breaks my mind, but the standard is decidedly mute on the subject, so I suppose that's just what the crystal folks picked when they wrote it.

Reply:
In most cases it's bad practice to have multiple params the same key anyway. One exception being selects with multiple values.

Reply:
In most cases it's bad practice to have multiple params the same key anyway. One exception being selects with multiple values.

Reply:
Sorry, was confused about the line we were talking about.